commands/command_lock.go

```diff
@@ -0,0 +1,89 @@
+package commands
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/github/git-lfs/api"
+	"github.com/github/git-lfs/config"
+	"github.com/github/git-lfs/git"
+	"github.com/spf13/cobra"
+)
+
+var (
+	lockRemote     string
+	lockRemoteHelp = "specify which remote to use when interacting with locks"
+
+	// TODO(taylor): consider making this (and the above flag) a property of
+	// some parent-command, or another similarly less ugly way of handling
+	// this
+	setLockRemoteFor = func(c *config.Configuration) {
+		c.CurrentRemote = lockRemote
+	}
+
+	lockCmd = &cobra.Command{
+		Use: "lock",
+		Run: lockCommand,
+	}
+)
+
+func lockCommand(cmd *cobra.Command, args []string) {
+	setLockRemoteFor(config.Config)
+
+	if len(args) == 0 {
+		Print("Usage: git lfs lock <path>")
+		return
+	}
+
+	latest, err := git.CurrentRemoteRef()
+	if err != nil {
+		Error(err.Error())
+		Exit("Unable to determine lastest remote ref for branch.")
+	}
+
+	path, err := lockPath(args[0])
+	if err != nil {
+		Error(err.Error())
+	}
+
+	s, resp := API.Locks.Lock(&api.LockRequest{
+		Path:               path,
+		Committer:          api.CurrentCommitter(),
+		LatestRemoteCommit: latest.Sha,
```

Comment:
Yeah, this is why I said there's branching cases both before and after the client requests the lock. In case 2 for example, the client will have to take action **before even requesting** the lock from the server, because it knows even before asking the server that the non-empty diff means there's a problem with obtaining a lock with the current working copy state. 

I covered some of the sequence I thought might be required, both pre-request (client), during request (server) and post-grant (client) in https://github.com/github/git-lfs/blob/master/docs/proposals/locking.md. But I agree that it's fine to concentrate on getting the basic locking API working on both ends first. I just wanted to flag up the conflict cases we're trying to address to make sure they can be dealt with. On reflection I think they can with this API, so long as the pre/post functionality is there. 

FWIW some other cases not covered above:

- Client tries to obtain a lock on a remote ref and thinks there's an empty diff, but server knows that client's record of the remote ref is out of date (behind) and there is a diff now.
- Client branch is tracking a remote ref and they both agree, but server has another remote ref which is a descendant which has a change to this file in it (lock and unlock happened separately). Granting the lock just based on this remote ref will cause a conflict later on.
- Client's tracked remote ref has a sibling branch on the server where changes have occurred to this file in parallel. Again even though the file is not locked, granting the lock will cause a conflict later. 

I talked about these cases in my proposal by suggesting the user would have to choose between switching branches, cherry-picking file state or accepting that they would have to resolve a conflict (discard one side) later. Some of these conditions have to be detected before requesting the lock (the server isn't aware of the working copy state so the client must act on any known issues there before requesting the lock), and some after (only the server knows the state of all remote branches).

We already know these cases exist so I don't think we need to wait until the community finds them, but at the same time they don't have to all be tackled right now. Just so long as we're aware of them this can be the next phase after the base API is working.